File: src/LIT.Smabu.UseCases/Catalogs/CatalogItemDTO.cs
﻿using LIT.Smabu.Domain.CatalogAggregate;
using LIT.Smabu.Domain.Common;
using LIT.Smabu.UseCases.Shared;

namespace LIT.Smabu.UseCases.Catalogs
{
    public record CatalogItemDTO(CatalogItemId Id, CatalogGroupId CatalogGroupId, CatalogId CatalogId, CatalogItemNumber Number, string Name, 
        string Description, bool IsActive, string GroupName,
        CatalogItemPrice[] Prices, CustomerCatalogItemPrice[] CustomerPrices, Unit Unit, CatalogItemPrice CurrentPrice) : IDTO
    {
        public string DisplayName => $"{Number.DisplayName} {Name}";
        public Currency Currency { get; } = Currency.EUR;

        public static CatalogItemDTO Create(CatalogItem item, CatalogGroup catalogGroup)
        {
            return new CatalogItemDTO(item.Id, item.CatalogGroupId, item.CatalogId, item.Number, item.Name, item.Description, item.IsActive, 
                catalogGroup.Name, [.. item.Prices], [.. item.CustomerPrices], item.Unit, item.GetCurrentPrice());
        }
    }
}

File: src/lit.smabu.react/src/types/domain/catalog-item-dto.ts

import { CatalogGroupId } from './catalog-group-id';
import { CatalogId } from './catalog-id';
import { CatalogItemId } from './catalog-item-id';
import { CatalogItemNumber } from './catalog-item-number';
import { CatalogItemPrice } from './catalog-item-price';
import { Currency } from './currency';
import { CustomerCatalogItemPrice } from './customer-catalog-item-price';
import { Unit } from './unit';

export interface CatalogItemDTO {
  id?: CatalogItemId;
  catalogGroupId?: CatalogGroupId;
  catalogId?: CatalogId;
  number?: CatalogItemNumber;
  name?: string | null;
  description?: string | null;
  isActive?: boolean;
  groupName?: string | null;
  prices?: Array<CatalogItemPrice> | null;
  customerPrices?: Array<CustomerCatalogItemPrice> | null;
  unit?: Unit;
  currentPrice?: CatalogItemPrice;
  displayName?: string | null;
  currency?: Currency;
}
